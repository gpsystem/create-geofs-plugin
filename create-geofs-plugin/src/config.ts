import { dirname, join, resolve } from "node:path";
import { cwd } from "node:process";
import { Command } from "commander";
import { version } from "../package.json";

/**
 * The processed configuration. This is the result of parsing the command line arguments.
 */
export interface Configuration {
  targetDir: string;
  templateDir: string;
  templateName: "default" | "typescript" | "react" | "react-ts";
  templateTopLevelDir: string;
  overwrite: boolean;
}

/**
 * Parses the command line arguments and returns the configuration.
 * @param argv The command line arguments. Should always be process.argv.
 * @returns The processed configuration.
 */
export function parseConfig(argv: string[]): Configuration {
  let destination = "";

  let { overwrite, template } = new Command()
    .name("create-geofs-plugin")
    .version(`Create GeoFS Plugin v${version}`)
    .argument("<destination>")
    .action((dest: string) => {
      // TODO: is this check a bit too harsh?
      if (!dest.includes("/"))
        throw new Error("argument <destination> must be a relative path");
      else destination = resolve(cwd(), dest);
    })
    .option("--overwrite", "Overwrite existing files", false)
    .option("-t, --template <template>", "Name of the template to use")
    .parse(argv)
    .opts<{ overwrite?: boolean; template?: string }>();

  if (!["default", "typescript", "react", "react-ts"].includes(template ?? ""))
    template = "default";

  // hack: getting the directory name of a file that will always be in the top-level
  const templateTopLevelDir: string = dirname(
    require.resolve(`@geps/cgp-template-${template}/README.md`)
  );

  return {
    overwrite: overwrite ?? false,
    templateDir: join(templateTopLevelDir, "template"),
    targetDir: destination,
    templateTopLevelDir,
    // template is validated against an array of the possible template names, so this assertion is safe
    templateName: template as "default" | "typescript" | "react" | "react-ts",
  };
}
