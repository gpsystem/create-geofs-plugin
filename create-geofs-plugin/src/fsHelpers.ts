import {
  Dirent,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";

function pathIsDirectory(path: string): boolean {
  return statSync(path).isDirectory();
}

function internalWriteFileSync(targetFile: string, data: string): void {
  writeFileSync(targetFile, data, "utf-8");
}

export function outputFile(
  targetFile: string,
  data: string,
  {
    force = false,
  }: {
    force?: boolean;
  } = {}
): void {
  if (!force && existsSync(targetFile)) throw new Error("file already exists");

  const targetFileDirectory: string = dirname(targetFile);

  if (!existsSync(targetFileDirectory))
    mkdirSync(targetFileDirectory, { recursive: true });
  internalWriteFileSync(targetFile, data);
}

export function* getAllFilesInDir(
  dirPath: string
): Generator<string, void, void> {
  if (!pathIsDirectory(dirPath))
    throw new Error(`passed path is not a directory`);

  const directoryDetails: Dirent[] = readdirSync(dirPath, {
    withFileTypes: true,
  });

  for (const directoryChild of directoryDetails) {
    const childPath: string = resolve(dirPath, directoryChild.name);

    if (directoryChild.isDirectory()) yield* getAllFilesInDir(childPath);
    else if (directoryChild.isFile()) yield childPath;
  }
}

export interface CopyDirOptions {
  overwrite?: boolean;
}

export function copyDir(
  srcDir: string,
  targetDir: string,
  { overwrite = false }: CopyDirOptions = {}
): void {
  if (srcDir === targetDir)
    throw new Error("source and target directories can't be the same");
  if (!pathIsDirectory(srcDir))
    throw new Error("source path isn't a directory");

  for (const srcFilePath of getAllFilesInDir(srcDir)) {
    const newFilePath: string = join(targetDir, relative(srcDir, srcFilePath));

    if (existsSync(newFilePath) && overwrite === false) continue;
    else
      outputFile(newFilePath, readFileSync(srcFilePath, "utf-8"), {
        force: overwrite,
      });
  }
}
