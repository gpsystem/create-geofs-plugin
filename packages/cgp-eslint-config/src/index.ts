import type { EslintConfig } from "eslint-define-config";
import merge from "lodash.merge";
import {
  eslintConfigBaseDependencies,
  eslintConfigBases,
  EslintConfigNames,
} from "./configurations";

type MinOneElementArray<T> = [T, ...T[]];

/**
 * Merges eslint configuration templates into one configuration to output to a .eslintrc file.
 *
 * @param configNames All the configuration templates.
 * The templates will be merged (deeply) with further right configuration templates overwriting further left templates.
 *
 * @returns A tuple. First is the configuration, ready to be stringified and outputted to a .eslintrc file.
 * Second is the names of the dependencies to install to make sure the .eslintrc configuration works properly.
 */
export function getEslintConfig(
  ...configNames: MinOneElementArray<EslintConfigNames>
): [config: EslintConfig, dependencies: string[]] {
  if (configNames.length < 1)
    throw new Error("must provide at least one config name");

  const dependencies: string[] = configNames.flatMap(
    (val) => eslintConfigBaseDependencies[val]
  );
  const configsToMerge = configNames.map<EslintConfig>(
    (val) => eslintConfigBases[val]
  ) as MinOneElementArray<EslintConfig>;
  const config: EslintConfig = merge(...configsToMerge);

  // if the typescript template is used, remove globals from the config (typescript-eslint will pick them up)
  if (configNames.includes("tsBase") && config.globals) delete config.globals;

  return [config, dependencies];
}

export { EslintConfigNames };
