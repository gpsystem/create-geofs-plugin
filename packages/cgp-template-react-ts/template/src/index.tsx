/// <reference types="jquery" />
/// <reference types="@geps/geofs-types" />

import React from "react";
import { createRoot } from "react-dom/client";
import { App } from "./App";

// Replace the selector with a link to your UI host!
// Example: #geofs-ui-right
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const root = createRoot(document.querySelector("REPLACE ME")!);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
