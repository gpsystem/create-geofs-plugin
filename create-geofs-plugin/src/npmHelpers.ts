import { execSync } from "node:child_process";
import { join } from "node:path";
import { outputFile } from "./fsHelpers";

export function initializeNpm(
  targetDir: string,
  templateDeps: string[],
  { ci = false }: { ci?: boolean } = {}
): void {
  let firstPackageJson: Record<string, unknown> = {
    name: "geofs-plugin",
    version: "",
    private: true,
    scripts: {
      build: "webpack",
      test: "npm run lint",
    },
  };

  if (ci)
    firstPackageJson = {
      ...firstPackageJson,
      templateDeps__INSTALLED_WHEN_NOT_IN_CI: templateDeps,
    };

  outputFile(
    join(targetDir, "package.json"),
    JSON.stringify(firstPackageJson, null, 2),
    {
      force: true,
    }
  );

  if (!ci) execSync(`npm i ${templateDeps.join(" ")} -D`, { cwd: targetDir });
}
