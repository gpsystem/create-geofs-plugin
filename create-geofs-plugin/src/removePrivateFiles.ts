import { rmSync } from "node:fs";
import { join, relative, sep } from "node:path";
import { getAllFilesInDir } from "./fsHelpers";

export default function removePrivateFiles(directoryPath: string): void {
  for (const fullFilePath of getAllFilesInDir(directoryPath)) {
    const relFilePathParts: string[] = relative(
      directoryPath,
      fullFilePath
    ).split(sep);

    for (const [i, pathPortion] of relFilePathParts.entries()) {
      // all private entries start with __
      if (pathPortion.startsWith("__")) {
        const relativePathPortion: string = relFilePathParts
          .slice(0, i + 1)
          .join("/");
        // keep in mind that fullPath can be a directory or file
        const fullPath: string = join(directoryPath, relativePathPortion);

        // fullPath could be a directory or a file
        rmSync(fullPath, { force: true, recursive: true });
        // if there are more parts to the path, they would have to be nested files or subdirectories
        // they would have been removed by the rmSync call, so skip over all of them
        break;
      }
    }
  }
}
