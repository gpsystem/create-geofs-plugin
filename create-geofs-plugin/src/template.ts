import { readFileSync } from "node:fs";
import { EslintConfigNames, getEslintConfig } from "@geps/cgp-eslint-config";

export interface RawTemplateJson {
  /** The dependencies of the template. Does not include the dependencies returned from cgp-eslint-config. */
  dependencies: string[];
  eslintConfigTemplates: EslintConfigNames[];
}

/** Checks if the input is a vanilla object. */
function isObject(obj: unknown): obj is Record<string, unknown> {
  // intentionally allowing a comparison of both undefined and null, since
  // Object.getPrototypeOf will throw for both of them
  // eslint-disable-next-line eqeqeq
  return obj != null && Object.getPrototypeOf(obj) === Object.prototype;
}

/** Checks if the provided input follows the format of a raw template.json file. */
export function assertTemplateJsonFormat(
  templateJson: unknown
): templateJson is RawTemplateJson {
  // asserts that allPossibleConfigNames is a valid array of all possible EslintConfigNames
  // thanks to Joshua Chen on the TS discord for this assertion code!
  type AssertTrue<T extends true> = T;
  // If typescript throws type 'false' does not satisfy the constraint 'true' here, then
  // allPossibleConfigNames is malformed
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  type A = AssertTrue<
    [EslintConfigNames] extends [typeof allPossibleConfigNames[number]]
      ? true
      : false
  >;

  const allPossibleConfigNames = ["baseConfig", "tsBase", "reactBase"] as const;
  const { isArray } = Array;

  return (
    isObject(templateJson) &&
    isArray(templateJson.dependencies) &&
    templateJson.dependencies.every<string>(
      (val): val is string => typeof val === "string"
    ) &&
    isArray(templateJson.eslintConfigTemplates) &&
    templateJson.eslintConfigTemplates.every<EslintConfigNames>(
      (val): val is EslintConfigNames => allPossibleConfigNames.includes(val)
    )
  );
}

export interface ExpandedTemplateJson {
  /** All dependencies of the template, including the eslint dependencies. */
  dependencies: string[];
  /** The configuration to save to .eslintrc.yml. */
  eslintConfig: Record<string, unknown>;
}

/**
 * Expands a raw template.json file into a data structure to be used.
 *
 * @param templateJsonPath The path to the raw template.json file.
 * @returns The expanded template.json file.
 */
export default function expandTemplateJson(
  templateJsonPath: string
): ExpandedTemplateJson {
  const templateJson: unknown = JSON.parse(
    readFileSync(templateJsonPath, "utf-8")
  );

  if (!assertTemplateJsonFormat(templateJson))
    throw new Error("template.json is malformed");

  const { eslintConfigTemplates } = templateJson;
  if (eslintConfigTemplates.length === 0)
    throw new Error("template must specify at least one config template");
  if (eslintConfigTemplates[0] !== "baseConfig")
    throw new Error(
      "template's eslint config template list must include the base configuration first"
    );

  const [eslintConfig, eslintDependencies] = getEslintConfig(
    ...(eslintConfigTemplates as [EslintConfigNames, ...EslintConfigNames[]])
  );

  const dependencies: string[] = [
    ...templateJson.dependencies,
    ...eslintDependencies,
  ];

  return {
    dependencies,
    // EslintConfig is an interface, which isn't assignable to an object at compile-time, but is an object at runtime
    eslintConfig: eslintConfig as Record<string, unknown>,
  };
}
