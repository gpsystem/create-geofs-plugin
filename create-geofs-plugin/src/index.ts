import { join } from "node:path";
import { dump } from "js-yaml";
import { dir, DirectoryResult } from "tmp-promise";
import { Configuration, parseConfig } from "./config";
import { copyDir, outputFile } from "./fsHelpers";
import { initializeNpm } from "./npmHelpers";
import removePrivateFiles from "./removePrivateFiles";
import expandTemplateJson, { ExpandedTemplateJson } from "./template";

export async function start(
  argv: string[],
  { ci = false }: { ci?: boolean } = {}
): Promise<void> {
  const config: Configuration = parseConfig(argv);
  const { path: tmpDirPath, cleanup: cleanupTempDir }: DirectoryResult =
    await dir({
      unsafeCleanup: true,
    });

  copyDir(config.templateDir, tmpDirPath);
  const { dependencies: templateDeps, eslintConfig }: ExpandedTemplateJson =
    expandTemplateJson(join(config.templateTopLevelDir, "template.json"));

  removePrivateFiles(tmpDirPath);
  outputFile(
    join(tmpDirPath, ".eslintrc.yml"),
    dump(eslintConfig, {
      sortKeys: true,
    })
  );
  copyDir(tmpDirPath, config.targetDir, { overwrite: config.overwrite });
  initializeNpm(config.targetDir, templateDeps, { ci });

  await cleanupTempDir();
}
