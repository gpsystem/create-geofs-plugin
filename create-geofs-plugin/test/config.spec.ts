import { relative } from "node:path";
import { afterEach, describe, expect, test } from "@jest/globals";
import { Configuration, parseConfig } from "../src/config";
import {
  allArgPermutations,
  createArgsForProgram,
  createDirectoryForTest,
  normalizeToForwardSlash,
} from "./utils/index";

describe("parse config", () => {
  const { cleanup, directoryPath, getRelativePath } = createDirectoryForTest();

  afterEach(() => cleanup());

  for (const [name, args] of allArgPermutations) {
    test(`ensure parsing integrity ${name}`, () => {
      const originalConfig: Configuration = parseConfig(
        createArgsForProgram(args, directoryPath)
      );
      const snapshotConfig: Configuration = {
        ...originalConfig,
        // should be an empty string
        targetDir: normalizeToForwardSlash(
          getRelativePath(originalConfig.targetDir)
        ),
        templateDir: normalizeToForwardSlash(
          relative(__dirname, originalConfig.templateDir)
        ),
        templateTopLevelDir: normalizeToForwardSlash(
          relative(__dirname, originalConfig.templateTopLevelDir)
        ),
      };

      expect(snapshotConfig).toMatchSnapshot();
    });
  }
});
