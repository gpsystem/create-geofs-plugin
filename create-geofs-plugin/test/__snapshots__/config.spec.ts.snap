// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse config ensure parsing integrity overwrite:false/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-default/template",
  "templateName": "default",
  "templateTopLevelDir": "../../packages/cgp-template-default",
}
`;

exports[`parse config ensure parsing integrity overwrite:false//template:react/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react/template",
  "templateName": "react",
  "templateTopLevelDir": "../../packages/cgp-template-react",
}
`;

exports[`parse config ensure parsing integrity overwrite:false//template:react-ts/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react-ts/template",
  "templateName": "react-ts",
  "templateTopLevelDir": "../../packages/cgp-template-react-ts",
}
`;

exports[`parse config ensure parsing integrity overwrite:false//template:typescript/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-typescript/template",
  "templateName": "typescript",
  "templateTopLevelDir": "../../packages/cgp-template-typescript",
}
`;

exports[`parse config ensure parsing integrity overwrite:true/ 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-default/template",
  "templateName": "default",
  "templateTopLevelDir": "../../packages/cgp-template-default",
}
`;

exports[`parse config ensure parsing integrity overwrite:true//template:react/ 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react/template",
  "templateName": "react",
  "templateTopLevelDir": "../../packages/cgp-template-react",
}
`;

exports[`parse config ensure parsing integrity overwrite:true//template:react-ts/ 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react-ts/template",
  "templateName": "react-ts",
  "templateTopLevelDir": "../../packages/cgp-template-react-ts",
}
`;

exports[`parse config ensure parsing integrity overwrite:true//template:typescript/ 1`] = `
Object {
  "overwrite": true,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-typescript/template",
  "templateName": "typescript",
  "templateTopLevelDir": "../../packages/cgp-template-typescript",
}
`;

exports[`parse config ensure parsing integrity template:react/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react/template",
  "templateName": "react",
  "templateTopLevelDir": "../../packages/cgp-template-react",
}
`;

exports[`parse config ensure parsing integrity template:react-ts/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-react-ts/template",
  "templateName": "react-ts",
  "templateTopLevelDir": "../../packages/cgp-template-react-ts",
}
`;

exports[`parse config ensure parsing integrity template:typescript/ 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-typescript/template",
  "templateName": "typescript",
  "templateTopLevelDir": "../../packages/cgp-template-typescript",
}
`;

exports[`parse config ensure parsing integrity vanilla 1`] = `
Object {
  "overwrite": false,
  "targetDir": "",
  "templateDir": "../../packages/cgp-template-default/template",
  "templateName": "default",
  "templateTopLevelDir": "../../packages/cgp-template-default",
}
`;
